### PR TITLE
MINT-748 Refactor voting contracts

### PR DESCRIPTION
**Pull request by alesanro**
[JIRA MINT-748](https://chronobank.atlassian.net/browse/MINT-748)

Implement a new set of contracts for voting to separate responsibilities and reduce the amount of gas that was spent for `vote` process

Use manager -> proxies scheme to reduce the size of contracts and optimize migration and redeploying expenses for changing functionality.

### CODE REVIEW

**Comment by alesanro** [comment]
@ahiatsevich Please, review.

File: contracts/vote/v2/PollEntityFactory.sol [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        contractsManager = _contractsManager;
+        backend = _backend;
+    }
+
+    /**
+    * @dev Creates a new poll and provides all needed data for its instantiation
+    *
+    * @return address of a brand new poll contract
+    */
+    function createPoll() public constant returns (address) {
```
Comment (ahiatsevich) [inline contracts/vote/v2/PollEntityFactory.sol]: Move constrctor's params to `createPoll`?

File: contracts/vote/v2/PollEntityRouter.sol [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    function destroy() onlyBackendOrVotingManager public {
+        selfdestruct(owner);
+    }
+
+    /**
+    * @dev Gets address of a backend contract
+    *
+    * @return _backend address of a backend contract
+    */
+    function backend() internal constant returns (address) {
```
Comment (ahiatsevich) [inline contracts/vote/v2/PollEntityRouter.sol]: backendAddress already internal?

File: contracts/vote/v2/VotingManager.sol [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+        }
+
+        _polls = new address[](_pageSize);
+        uint _lastIndex = _startIndex + _pageSize;
+        _lastIndex = (_lastIndex >= _pollsCount) ? _pollsCount : _lastIndex;
+        for (uint _idx = _startIndex; _idx < _lastIndex; ++_idx) {
+            _polls[_idx] = store.get(pollsStorage, _idx);
+        }
+
+        _nextIndex = _lastIndex + 1;
```
Comment (ahiatsevich) [inline contracts/vote/v2/VotingManager.sol]: TODO: nextIndex?

File: contracts/vote/v2/VotingManager.sol [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+    * @param _deadline time after which poll isn't active anymore
+    *
+    * @return (poll address, resultCode). OK if all went all right, error code otherwise
+    */
+    function createPoll(bytes32[16] _options, bytes32[4] _ipfsHashes, bytes32 _detailsIpfsHash, uint _votelimit, uint _deadline) public returns (address, uint) {
+        PollEntityFactoryInterface _pollsFactory = PollEntityFactoryInterface(store.get(pollsFactoryStorage));
+
+        address _entity = _pollsFactory.createPoll();
+        uint _resultCode = PollEntityInterface(_entity).init(_options, _ipfsHashes, _detailsIpfsHash, _votelimit, _deadline, msg.sender);
+        if (_resultCode != OK) {
```
Comment (ahiatsevich) [inline contracts/vote/v2/VotingManager.sol]: return result code when changes has been performed

File: contracts/vote/v2/VotingManager.sol [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+    /**
+    * @dev DO NOT СALL IT DIRECTLY. Used by a poll contract.
+    * Emits PollVoted event in case of successful voting.
+    *
+    * @param _user address of a user who votes
+    * @param _choice option chosen by user
+    *
+    * @return result code of an operation
+    */
+    function vote(address _user, uint8 _choice) onlyPoll public returns (uint) {
```
Comment (ahiatsevich) [inline contracts/vote/v2/VotingManager.sol]: set events history to polls(or better = lookup via contracts manager)

**Pull request merged**